Add GIF support for animated lock screen

sh1r4s3 opened the pull request:
## Description
Use gif_lib to read a [GIF file](https://www.w3.org/Graphics/GIF/spec-gif89a.txt) and show animated lock screen.
Currently it supports only 0-2 disposal methods.

### Screenshots/screencaps
![demonstation](https://user-images.githubusercontent.com/21184954/174704230-72f919ad-05d1-4b07-bb0b-f9884675c409.gif)

## Release notes
Added support for animated GIF images.

sh1r4s3 committed:
Add GIF support for animated lock screen

Use gif_lib to read a GIF file [1] and show animated lock screen.
Currently it supports only 0-2 disposal methods.

[1] https://www.w3.org/Graphics/GIF/spec-gif89a.txt

Raymo111 commented on i3lock.c:
```diff
@@ -61,6 +61,9 @@
 #include "fonts.h"
 
+#include <gif_lib.h>
+#include <arpa/inet.h>
```
Why is this necessary?

sh1r4s3 replied:
@Raymo111 thanks for the comment. Yes, `#include <arpa/inet.h>` is not needed. I wanted to use it for byteorder(3) functions, but in the result I didn't use them. I updated the PR.

Raymo111 replied:
Ok.

Raymo111 commented:
@sh1r4s3 Can you fix the build files as well?

sh1r4s3 committed:
Add libgif-dev to Dockerfile and check for gif_lib.h in configure

sh1r4s3 commented:
@Raymo111 I updated the Dockerfile to include `libgif-dev` package that should contain `gif_lib.h` as well as the shared lib. Also I updated `configure.ac` to check that `gif_lib.h` exists. Could you please trigger the workflow to test those changes?

sh1r4s3 commented:
@Raymo111 From the build log it is seems that it doesn't use an updated version of `travis/Dockerfile`. Not sure why... Maybe it fetches this Dockerfile from the different branch or just uses the cached image.

Raymo111 commented:
@sh1r4s3 oh we're not using Travis... Upstream used to use it but they switched to GitHub Actions recently and we've been using GitHub Actions for a while now. You'll want to edit the `.github/` files

sh1r4s3 commented:
@Raymo111 got it, thanks! I updated these files. Could you please trigger the build again :)

sh1r4s3 committed:
Fix CodeQL error

Multiplication result converted to larger type

sh1r4s3 commented:
@Raymo111 seems that now everything is OK :)

Raymo111 commented on i3lock.c:
```diff
@@ -1225,36 +1369,81 @@ static cairo_surface_t *read_raw_image(const char *image_path, const char *image
     // https://www.w3.org/TR/2003/REC-PNG-20031110/#5PNG-file-signature
     static unsigned char PNG_REFERENCE_HEADER[8] = {137, 80, 78, 71, 13, 10, 26, 10};
     if (memcmp(PNG_REFERENCE_HEADER, png_header, sizeof(png_header)) != 0) {
-        DEBUG("File \"%s\" does not start with a PNG header. i3lock currently only supports loading PNG files.\n", image_path);
```
Good catch!

Raymo111 commented on i3lock.c:
```diff
@@ -1225,36 +1369,81 @@ static cairo_surface_t *read_raw_image(const char *image_path, const char *image
+    }
+
+    /* Check file exists and has correct header */
+    FILE *file = fopen(image_path, "r");
```
Any specific reason `rb` from the original was switched to `r`?

sh1r4s3 replied:
No, but it would be better to use `rb` for portability. Thanks!

Raymo111 commented on i3lock.c:
```diff
@@ -1225,36 +1369,81 @@ static cairo_surface_t *read_raw_image(const char *image_path, const char *image
+    /* Check file exists and has correct header */
+    FILE *file = fopen(image_path, "r");
+    if (file == NULL) {
+        fprintf(stderr, "Image file path \"%s\" cannot be opened: %s\n", image_path, strerror(errno));
```
Say `...opened (Error %s)\n` instead

Raymo111 commented on i3lock.c:
```diff
@@ -1446,31 +1635,41 @@ static void raise_loop(xcb_window_t window) {
-        unsigned char* jpg_data = read_JPEG_file(image_path, &jpg_info);
+    unsigned char *jpg_data;
+
+    /* Create a pixmap to render on, fill it with the background color */
```
This comment isn't supposed to be here, is it?

sh1r4s3 replied:
Yeah.. it was an accident :)

Raymo111 commented on i3lock.c:
```diff
@@ -1446,31 +1635,41 @@ static void raise_loop(xcb_window_t window) {
+            /* Create a pixmap to render on, fill it with the background color */
+            img = cairo_image_surface_create_from_png(image_path);
+            break;
+        case IMAGE_FORMAT_GIF:
```
Move GIF to after JPEG so we put static images together

Raymo111 commented on i3lock.c:
```diff
@@ -1446,31 +1635,41 @@ static void raise_loop(xcb_window_t window) {
             }
+            break;
+        default:
+            fprintf(stderr, "Image format of the file %s is unknown\n", image_path);
```
Change this message to `Unsupported image file format: %s\n`

Raymo111 commented on i3lock.c:
```diff
@@ -1526,6 +1725,17 @@ bool load_slideshow_images(const char *path) {
+
+    if (++img_count >= gif_img_count) img_count = 0;
+    img = gif_img[img_count].img;
+    redraw_screen();
```
Will this cause issues on very-high-framerate GIFs?

sh1r4s3 replied:
It might cause issues in that case, I believe. But the rate should be quite high and it depends on hardware. I thought that we can skip some frames if the rate greater than 1/60 but then I remembered that in GIF the delays between frames can vary and independent of each other. Not sure if there is an easy way to skip frames for that case..

Raymo111 commented on i3lock.c:
```diff
@@ -1526,6 +1725,17 @@ bool load_slideshow_images(const char *path) {
+    if (++img_count >= gif_img_count) img_count = 0;
+    img = gif_img[img_count].img;
+    redraw_screen();
+    ev_timer_stop(loop, timer);
```
If `redraw_screen` takes non-negligible time, should this be run before or after it?

sh1r4s3 replied:
In that particular case, as I understand, it doesn't matter. It is because we need to restart the timer explicitly each time we enter into the callback. But, probably, it would be better to stop the timer before redraw as it confuse less when reading the code.

Raymo111 commented on i3lock.c:
```diff
@@ -247,7 +249,16 @@ char *image_path = NULL;
+int gif_img_count = 0;
+
 cairo_surface_t *img = NULL;
+bool tile = false;
```
Is this variable ever used?

sh1r4s3 replied:
No.. it has slipped through! Removed :)

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+    /* Open and load a GIF file */
+    GifFileType *gif = DGifOpenFileName(image_path, &err);
+    if (!gif) {
+        fprintf(stderr, "Could not open GIF file, err=%d\n", err);
```
Use this instead: `Could not open GIF file, (Error %d)\n`

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+        return NULL;
+    }
+    if (DGifSlurp(gif) != GIF_OK) {
+        printf("Could not read the GIF image\n");
```
Should be printed to `stderr`, no?

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+    cmap = gif->SColorMap;
+    bg_idx = gif->SBackGroundColor;
+    GifColorType *cmap_bg_rgb = cmap->Colors + bg_idx;
+    bg_color = 0 | cmap_bg_rgb->Blue | cmap_bg_rgb->Green<<8 | cmap_bg_rgb->Red<<16;
```
Is the initial 0 needed? Also, spaces around bitshift operators please

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+    bg_idx = gif->SBackGroundColor;
+    GifColorType *cmap_bg_rgb = cmap->Colors + bg_idx;
+    bg_color = 0 | cmap_bg_rgb->Blue | cmap_bg_rgb->Green<<8 | cmap_bg_rgb->Red<<16;
+    stride = cairo_format_stride_for_width(CAIRO_FORMAT_RGB24, width)/sizeof(uint32_t);
```
Space on either side of the `/` please

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+    bg_color = 0 | cmap_bg_rgb->Blue | cmap_bg_rgb->Green<<8 | cmap_bg_rgb->Red<<16;
+    stride = cairo_format_stride_for_width(CAIRO_FORMAT_RGB24, width)/sizeof(uint32_t);
+    if (stride < 0) {
+        fprintf(stderr, "Invalid stride\n");
```
What's a stride? Have the error message be more user-friendly :)

sh1r4s3 replied:
Basically, the [stride](https://www.cairographics.org/manual/cairo-Image-Surfaces.html#cairo-image-surface-get-stride) is the real size of a row, because a row can be padded. From the link:
> The stride is the distance in bytes from the beginning of one row of the image data to the beginning of the next row.

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+    gif_img = malloc(gif->ImageCount * sizeof(struct gif));
+    gif_img_count = gif->ImageCount;
+    if (!gif_img) {
+        fprintf(stderr, "Can't allocate memory for GIF image buffers\n");
```
Can't allocate -> Could not allocate

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+
+read_gif_image_clean:
+    if (!DGifCloseFile(gif, &err)) {
+        fprintf(stderr, "DGifCloseFile failed with err=%d\n", err);
```
I think this can be a `DEBUG` call instead, with string: `DGifCloseFile call failed (Error %d)\n`

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+
+                if (color_idx != gc.TransparentColor) {
+                    GifColorType *cmap_rgb = cmap_img->Colors + color_idx;
+                    uint32_t rgb = 0 | cmap_rgb->Blue | cmap_rgb->Green<<8 | cmap_rgb->Red<<16;
```
Is the initial 0 needed? Also, spaces around bitshift operators please

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+                if (x < img_left || x >= img_right || y < img_top || y >= img_bottom) {
+                    color_idx = bg_idx;
+                } else {
+                    color_idx = *(pimg->RasterBits + (x - img_left) + ((y - img_top)*img_width));
```
Spaces around multiplication operator please

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+        int img_bottom = img_top + pimg->ImageDesc.Height;
+
+        /* Handle disposal mode */
+        int data_size = width*height*((int)sizeof(width));
```
Spaces around multiplication operators please

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+        cairo_surface_flush(gif_img[idx].img);
+        uint32_t *data = (uint32_t *)cairo_image_surface_get_data(gif_img[idx].img);
+
+        // TODO: Chek L,R,W,H && color map?
```
Chek -> Check

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+        uint32_t *data = (uint32_t *)cairo_image_surface_get_data(gif_img[idx].img);
+
+        // TODO: Chek L,R,W,H && color map?
+        /* Find Graphics Control extension and delay */
```
Not quite sure what this does, can you explain?

sh1r4s3 replied:
The [GIF89 standard (paragraph 23)](https://www.w3.org/Graphics/GIF/spec-gif89a.txt) introduces *Graphic Control Extension*. Basically, this is an optional block that contains index for transparent color, delay and some other info for the particular frame (image). We need to read this block for each image packed in a GIF container.

Raymo111 commented on i3lock.c:
```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
+            switch (pext->Function) {
+                case GRAPHICS_EXT_FUNC_CODE:
+                    DGifExtensionToGCB(pext->ByteCount, pext->Bytes, &gc);
+                    gif_img[idx].delay_sec = gc.DelayTime*0.01;
```
Spaces around multiplication operator please. Also, why are we multiplying by 0.01?

sh1r4s3 replied:
`DelayTime` is in [ms], so we need to convert it to [s]. Added a comment about it.

Raymo111 replied:
Spaces

Raymo111 reviewed (changes requested):
Some changes requested. Also, mind explaining what `read_gif_image` does? I'm really not sure. Lastly, please update the manpage as well to document that GIFs are now supported (update the date to Sept 2022 as well). Great work and thanks for contributing, keep it up!

Raymo111 commented:
> Currently it supports only 0-2 disposal methods.

Btw @sh1r4s3 What did you mean by this?

sh1r4s3 committed:
Add info about GIF and JPG to the man page

sh1r4s3 commented:
@Raymo111 Thanks for the review! I implemented your suggestions and updated the man page.
> Also, mind explaining what `read_gif_image` does?

Sure. Basically it do the following:

1. Open a GIF file and read it to the buffer. gif_lib would parse the header so we can now extract some info like width/height of the canvas.
2. Next we want to calculate the [stride value](https://www.cairographics.org/manual/cairo-Image-Surfaces.html#cairo-image-surface-get-stride) which the cairo lib will use. The stride is the distance in bytes from the beginning of one row of the image data to the beginning of the next row.
3. The next step is to read frames (images) from the GIF file. In the loop we do:
3.1. Create a cairo surface.
3.2. Read the [Graphics Control Extension (paragraph .23)](https://www.w3.org/Graphics/GIF/spec-gif89a.txt) to extract the delay value. The delay value is the time in [ms] before switch to the next frame.
3.3. Next we read the size of the frame's image and the color map.
3.4. Read the disposal mode. Generally in GIF, currently, we have 4 disposal modes: 0 - No disposal specified; 1 - Do not dispose; 2 - Restore to background color; 3 - Restore to previous. Usually, AFAIK, only 0-2 are used. What does it means, e.g., we have 1 (Do not dispose) and we have some animation. The first frame is a background of the animation. The second frame contains mostly transparent pixels and a part of the rain drops, it would be written "on top" (basically replace pixels from the previous background image) of the previous image. The next frame contains the same, but the rain drops are in the new positions and on the places of the "old" rain drops there will be parts of the background image. And so on. The reason why I didn't implement the third disposal mode is that it uses very rarely and it is a little bit tricky to implement.

sh1r4s3 commented:
Also I think it is worth to mention that this code doesn't upscale GIF images. A user should find a gif file that suits its display's resolution or upscale it with the of other utility, like `convert` from ImageMagick (the way I use this code).

sh1r4s3 commented:
Hi @Raymo111, I just wanted to kindly remind about this PR :)

sh1r4s3 committed:
Update GH upload-artifact to v4 as v2 is deprecated

Raymo111 commented on i3lock.1:
```diff
@@ -8,7 +8,7 @@
 ..
 
-.TH i3lock-color 1 "JAN 2022" Linux "User Manuals"
+.TH i3lock-color 1 "SEP 2022" Linux "User Manuals"
```
Let's update this date 😄

sh1r4s3 committed:
Surround * with spaces

sh1r4s3 committed:
Update date in the man-page

sh1r4s3 commented:
Thanks man! :fire:

Raymo111 commented:
> Thanks man! 🔥

Haha np, thank *you* for persevering on this for 3 years!

ruudud commented:
This is awesome! Started watching this PR two years ago. Kudos for your perseverance @sh1r4s3 and thank you @Raymo111 for still maintaining this project ❤
